components/builder-depot/src/server.rs

```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+    let params = match validate_params(req, &["origin"]) {
+        Ok(p) => p,
+        Err(st) => return Ok(Response::with(st)),
+    };
+
+    let origin = &params["origin"];
+
+    let body = match req.get::<bodyparser::Struct<OriginSecretPayload>>() {
+        Ok(Some(body)) => {
+            if body.name.len() <= 0 {
+                return Ok(Response::with((
+                    status::UnprocessableEntity,
+                    "Missing value for field: `name`",
+                )));
+            }
+            if body.value.len() <= 0 {
+                return Ok(Response::with((
+                    status::UnprocessableEntity,
+                    "Missing value for field: `value`",
+                )));
+            }
+            body
+        }
+        _ => return Ok(Response::with(status::UnprocessableEntity)),
+    };
+
+    // get metadata from secret
+    let secret_metadata = match BoxKeyPair::secret_metadata(body.value.as_bytes()) {
+        Ok(res) => res,
+        Err(e) => {
+            return Ok(Response::with((
+                status::UnprocessableEntity,
+                format!("Failed to get metadata from payload: {}", e),
+            )))
+        }
+    };
+
+    debug!("Secret Metadata: {:?}", secret_metadata);
+
+    // fetch the private origin encryption key from the database
+    let mut db_priv_request = OriginPrivateEncryptionKeyGet::new();
+    match helpers::get_origin(req, origin) {
+        Ok(mut origin) => {
+            db_priv_request.set_owner_id(origin.get_owner_id());
+            db_priv_request.set_origin(origin.take_name());
+        }
+        Err(err) => return Ok(render_net_error(&err)),
+    }
+    let priv_key = match route_message::<
+        OriginPrivateEncryptionKeyGet,
+        OriginPrivateEncryptionKey,
+    >(req, &db_priv_request) {
+        Ok(key) => {
+            let key_str = from_utf8(key.get_body()).unwrap();
+            match BoxKeyPair::secret_key_from_str(key_str) {
+                Ok(key) => key,
+                Err(e) => {
+                    return Ok(Response::with(
+                        (status::UnprocessableEntity, format!("{}", e)),
+                    ))
+                }
+            }
+        }
+        Err(err) => return Ok(render_net_error(&err)),
+    };
+
+    let (name, rev) = match parse_name_with_rev(secret_metadata.sender) {
+        Ok(val) => val,
+        Err(e) => {
+            return Ok(Response::with((
+                status::UnprocessableEntity,
+                format!("Failed to parse name and revision: {}", e),
+            )))
+        }
+    };
+
+    debug!("Using key {:?}-{:?}", name, &rev);
+
+    // fetch the public origin encryption key from the database
+    let mut db_pub_request = OriginPublicEncryptionKeyGet::new();
+    match helpers::get_origin(req, origin) {
+        Ok(mut origin) => {
+            db_pub_request.set_owner_id(origin.get_owner_id());
+            db_pub_request.set_origin(origin.take_name());
```

Comment:
Is it worth caching these values in local variables from the call to `helpers::get_origin` above to avoid another network round trip?